File: C++/AESHelper.cpp
#include "AESHelper.h"
#include <vector>
#include <iostream>
#include <iterator>
#include <iomanip>
#include <stdexcept>
#include "Base64Helper.h"
#include "PBKDF2Helper.h"
#include "HashHelper.h"
#include "aes.h"
#include "modes.h"
#include "files.h"

bool AESHelper::DecryptDataPBKDF2(const std::string& data, const std::string& pbkdf2Password, const std::string& pbkdf2Salt, unsigned int pbkdf2Iterations, std::string& decryptedData)
{
	std::cout << "AES decryption of data started" << std::endl;

	if (pbkdf2Password.length() > 0 && pbkdf2Salt.length() > 0 && pbkdf2Iterations > 0)
	{
		// data and salt are base 64 encoded
		std::vector<byte> decodedPrivateKeyBytes;
		Base64Helper::Decode(data, decodedPrivateKeyBytes);

		std::vector<byte> decodedSalt;
		Base64Helper::Decode(pbkdf2Salt, decodedSalt);

		// derive bytestream from password and salt
		// via PBKDF2 - the resulting bytes (64) are 
		// two AES-256 keys which will be used in further steps
		PBKDF2Helper pbkdf2(pbkdf2Password, decodedSalt, pbkdf2Iterations);
		std::vector<byte> hashBytes;
		pbkdf2.GetBytes(64, hashBytes);
		auto cryptoKey = std::vector<byte>(hashBytes.begin(), hashBytes.begin() + 32);
		auto hmacKey = std::vector<byte>(hashBytes.begin() + 32, hashBytes.end());

		// the encrypted data holds an initialization vector
		// for the AES decryption, a HMAC-SHA-256 hash to
		// verify the given input and the actual private key bytes
		auto IVec = std::vector<byte>(decodedPrivateKeyBytes.begin(), decodedPrivateKeyBytes.begin() + 16);
		auto givenHmacHash = std::vector<byte>(decodedPrivateKeyBytes.begin() + 16, decodedPrivateKeyBytes.begin() + 48);
		auto privateKeyBytes = std::vector<byte>(decodedPrivateKeyBytes.begin() + 48, decodedPrivateKeyBytes.end());

		// it is necessary to compute the HMAC-SHA-256 hash
		// again to make sure the private key, password, salt
		// and iteration count weren't tampered with
		std::vector<byte> computedHmacHash;
		HashHelper::ComputeSHA256HMAC(privateKeyBytes, hmacKey, computedHmacHash);
		if (givenHmacHash != computedHmacHash)
		{
			throw std::runtime_error("HMAC hashes do not match, make sure you used a matching .bckey file and password");
		}

		AESHelper::DecryptData(privateKeyBytes, cryptoKey, IVec, decryptedData);

		std::cout << "AES decryption finished" << std::endl;
		return true;
	}
	else
	{
		throw std::runtime_error("Password and salt for the PBKDF2 algorithm can not be empty and the iteration count must be bigger than zero");
	}
}

bool AESHelper::DecryptFile(
	const std::string& encryptedFilePath, const std::vector<byte>& fileCryptoKey,
	const std::string& baseIVec, unsigned int blockSize, unsigned int offset,
	unsigned int padding, std::vector<byte>& decryptedFileBytes)
{
	std::cout << "AES decryption of file '" << encryptedFilePath << "' started" << std::endl;

	if (fileCryptoKey.size() > 0 && blockSize > 0)
	{
		// open the encrypted file, ...
		std::ifstream ifs(encryptedFilePath, std::ios::binary | std::ios::ate);
		if (!ifs.good())
		{
			std::string errorMsg("Encrypted file (" + encryptedFilePath + ") can't be opened (make sure the provided path is correct, the file exists and you have the right to open the file)");
			throw std::runtime_error(errorMsg.c_str());
		}

		// ... get the size ...
		std::streamoff pos = ifs.tellg();
		ifs.seekg(0, std::ios::beg);

		// ... and copy it into a byte vector to work with the data
		std::vector<char> fileBytes(static_cast<size_t>(pos));
		ifs.read(&fileBytes[0], pos);

		// IVec in file header is base 64 encoded
		std::vector<byte> decodedFileIV;
		Base64Helper::Decode(baseIVec, decodedFileIV);

		// create result vector and reserve enough space
		// to fit the complete decrypted file
		decryptedFileBytes.clear();
		decryptedFileBytes.reserve(fileBytes.size());
		unsigned long blockNo = 0;

		// report initial status
		size_t fileSize = fileBytes.size();
		std::string fileSizeStr = std::to_string(fileSize);
		auto fileSizeFivePer = static_cast<size_t>(std::floor(fileSize * 0.05));
		std::string byteProgress = " (0 / " + fileSizeStr + " bytes)";
		std::cout << "Progress: [" << std::setfill(' ') << std::setw(21) << "]" << std::left << std::setw(79) << byteProgress << std::right;

		// decrypt each block seperately with its own initialization vector
		for (size_t byteNo = offset, nextStatusThreshold = offset, currentStep = 0; byteNo < fileSize; byteNo += blockSize, ++blockNo)
		{
			auto blockIVec = AESHelper::ComputeBlockIVec(decodedFileIV, blockNo, fileCryptoKey);

			// get the input data for the current block (the last block may be shorter than [blockSize] bytes)
			auto iterEnd = (byteNo + blockSize >= fileSize) ? fileBytes.end() : fileBytes.begin() + byteNo + blockSize;
			std::vector<byte> blockInput(fileBytes.begin() + byteNo, iterEnd);

			// use PKCS7 padding for the last block if a cipher padding size greater than 0 was specified in file header
			auto currentPadding = (iterEnd == fileBytes.end() && padding > 0) ? CryptoPP::StreamTransformationFilter::PKCS_PADDING : CryptoPP::StreamTransformationFilter::NO_PADDING;

			// get the decrypted data for this block ...
			std::string decryptedBlockBytes;
			AESHelper::DecryptData(blockInput, fileCryptoKey, blockIVec, decryptedBlockBytes, currentPadding);

			// ... and append it to the data of previous blocks
			decryptedFileBytes.insert(decryptedFileBytes.end(), decryptedBlockBytes.begin(), decryptedBlockBytes.end());

			// report intermediate status every 5%
			if (byteNo > nextStatusThreshold)
			{
				int steps = byteNo / nextStatusThreshold;
				nextStatusThreshold += fileSizeFivePer * steps;

				currentStep += steps;
				byteProgress = " (" + std::to_string(byteNo) + " / " + fileSizeStr + " bytes)";
				std::cout << std::setfill('\b') << std::setw(100) << "" << std::setfill('#') << std::setw(currentStep) << "" << std::setfill(' ') << std::setw(21 - currentStep)
					<< "]" << std::left << std::setw(79) << byteProgress << std::right;
			}
		}

		// newline and buffer flush after status report
		byteProgress = " (" + fileSizeStr + " / " + fileSizeStr + " bytes)";
		std::cout << std::setfill('\b') << std::setw(100) << "" << std::setfill('#') << std::setw(21) << "]" << std::setfill(' ')
			<< std::left << std::setw(79) << byteProgress << std::right << std::endl;
		
		std::cout << "AES decryption of file finished" << std::endl;
		return true;
	}
	else
	{
		throw std::runtime_error("Crypto key for file can't be empty and block size must be bigger than zero");
	}
}

// from https://github.com/vgough/encfs/blob/559c30d01ed0a3d19258b12f15eae8785accc60f/encfs/SSL_Cipher.cpp#L626
/*private*/ std::vector<byte> AESHelper::ComputeBlockIVec(std::vector<byte> IVec, unsigned long seed, std::vector<byte> key)
{
	if (IVec.size() > 0 && key.size() > 0)
	{
		std::vector<byte> md(8);
		for (int i = 0; i < 8; ++i)
		{
			md.at(i) = static_cast<byte>(seed & 0xff);
			seed >>= 8;
		}

		std::vector<byte> data(IVec.size() + md.size());
		for (size_t i = 0; i < IVec.size(); ++i)
		{
			data.at(i) = IVec.at(i);
		}

		for (size_t i = 0; i < md.size(); ++i)
		{
			data.at(i + IVec.size()) = md.at(i);
		}

		std::vector<byte> buffer;
		HashHelper::ComputeSHA256HMAC(data, key, buffer, true);

		std::vector<byte> result(IVec.size());
		for (size_t i = 0; i < IVec.size(); ++i)
		{
			result.at(i) = buffer.at(i);
		}

		return result;
	}
	else
	{
		throw std::runtime_error("Base initialization vector and crypto key can't be empty");
	}
}

/*private*/ bool AESHelper::DecryptData(
	const std::vector<byte>& data, const std::vector<byte>& cryptoKey,
	const std::vector<byte>& IVec, std::string& output,
	CryptoPP::BlockPaddingSchemeDef::BlockPaddingScheme paddingMode /* = CryptoPP::StreamTransformationFilter::PKCS_PADDING*/)
{
	if (data.size() > 0 && cryptoKey.size() > 0 && IVec.size() > 0)
	{
		CryptoPP::CBC_Mode<CryptoPP::AES>::Decryption aesDecryptor;

		// provide the Crypto++ decryptor with the necessary
		// key and initialization vector needed for decryption
		const byte *key = cryptoKey.data();
		const byte *iv = IVec.data();
		aesDecryptor.SetKeyWithIV(key, cryptoKey.size(), iv, IVec.size());

		// Crypto++ takes the data (input) from the 'ArraySource' and transforms
		// it via the aesDecryptor into the decrypted output, which is then dumped
		// into the 'StringSink' (decryptedData)
		// PKCS7 padding (https://en.wikipedia.org/wiki/PKCS) is used in case the 
		// last data block is smaller than the block size used by AES (16 bytes)
		const byte *input = data.data();
		CryptoPP::ArraySource(
			input, data.size(), true,
			new CryptoPP::StreamTransformationFilter(
				aesDecryptor, new CryptoPP::StringSink(output), paddingMode));

		return true;
	}
	else
	{
		throw std::runtime_error("Encrypted data, crypto key and initialization vector can't be empty");
	}
}

File: Java/AESHelper.java
import org.bouncycastle.jce.provider.BouncyCastleProvider;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.Security;
import java.util.Arrays;
import java.util.stream.IntStream;

class AESHelper {
    static {
        Security.addProvider(new BouncyCastleProvider());
    }

    private static byte[] computeBlockIVec(final byte[] IVec, long seed, final byte[] key) {
        if (IVec.length <= 0 || key.length <= 0) {
            throw new RuntimeException("Base initialization vector and crypto key can't be empty");
        }

        byte[] md = new byte[8];
        for (int i = 0; i < 8; ++i) {
            md[i] = (byte) (seed & 0xff);
            seed >>= 8;
        }

        byte[] data = new byte[IVec.length + md.length];
        System.arraycopy(IVec, 0, data, 0, IVec.length);
        System.arraycopy(md, 0, data, IVec.length, md.length);

        byte[] buffer = HashHelper.ComputeSHA256HMAC(data, key, true);
        byte[] result = Arrays.copyOf(buffer, IVec.length);

        return result;
    }

    private static byte[] decryptData(final byte[] data, final byte[] cryptoKey, final byte[] IVec, Boolean isUserGeneratedData) throws UnsupportedEncodingException {
        return AESHelper.decryptData(data, cryptoKey, IVec, isUserGeneratedData, "PKCS5PADDING");
    }

    private static byte[] decryptData(final byte[] data, final byte[] cryptoKey, final byte[] IVec, Boolean isUserGeneratedData, String padding) throws UnsupportedEncodingException {
        if ((isUserGeneratedData ? data.length < 0 : data.length <= 0) || cryptoKey.length <= 0 || IVec.length <= 0) {
            throw new RuntimeException("Encrypted data, crypto key and initialization vector can't be empty");
        }

        byte[] result;
        try {
            // PKCS7 padding (https://en.wikipedia.org/wiki/PKCS) is used in case the
            // last data block is smaller than the block size used by AES (16 bytes)
            IvParameterSpec ivParamSpec = new IvParameterSpec(IVec);
            SecretKeySpec keySpec = new SecretKeySpec(cryptoKey, "AES");
            Cipher cipher = Cipher.getInstance("AES/CBC/" + padding);
            cipher.init(Cipher.DECRYPT_MODE, keySpec, ivParamSpec);
            result = cipher.doFinal(data);
        } catch (NoSuchAlgorithmException | NoSuchPaddingException
                | InvalidKeyException | InvalidAlgorithmParameterException
                | IllegalBlockSizeException | BadPaddingException e) {
            throw new RuntimeException("Data could not be decrypted", e);
        }

        return result;
    }

    public static byte[] decryptDataPBKDF2(
            final String data,
            final String pbkdf2Password,
            final String pbkdf2Salt,
            final int pbkdf2Iterations) throws UnsupportedEncodingException {
        System.out.println("AES decryption of data started");

        if (pbkdf2Password.isEmpty() || pbkdf2Salt.isEmpty() || pbkdf2Iterations <= 0) {
            throw new RuntimeException("Password and salt for the PBKDF2 algorithm can not be empty and the iteration count must be bigger than zero");
        }

        // data and salt are base 64 encoded
        byte[] decodedPrivateKeyBytes = Base64Helper.decode(data);
        byte[] decodedSalt = Base64Helper.decode(pbkdf2Salt);

        // derive bytestream from password and salt
        // via PBKDF2 - the resulting bytes (64) are
        // two AES256 keys which will be used in further steps
        PBKDF2Helper pbkdf2 = new PBKDF2Helper(pbkdf2Password, decodedSalt, pbkdf2Iterations);
        byte[] hashBytes = pbkdf2.getBytes(64);
        int[] unsignedHashBytes = new int[64];
        for (int i = 0; i < hashBytes.length; ++i) {
            unsignedHashBytes[i] = hashBytes[i] & 0xFF;
        }

        byte[] cryptoKey = Arrays.copyOfRange(hashBytes, 0, 32);
        byte[] hmacKey = Arrays.copyOfRange(hashBytes, 32, hashBytes.length);

        // the encrypted data holds an initialization vector
        // for the AES decryption, a HMAC-SHA-256 hash to
        // verify the given input and the actual private key bytes
        byte[] IVec = Arrays.copyOfRange(decodedPrivateKeyBytes, 0, 16);
        byte[] givenHmacHash = Arrays.copyOfRange(decodedPrivateKeyBytes, 16, 48);
        byte[] privateKeyBytes = Arrays.copyOfRange(decodedPrivateKeyBytes, 48, decodedPrivateKeyBytes.length);

        // it is necessary to compute the HMAC-SHA-256 hash
        // again to make sure the private key, password, salt
        // and iteraton count weren't tampered with
        byte[] computedHmacHash = HashHelper.ComputeSHA256HMAC(privateKeyBytes, hmacKey);
        if (!Arrays.equals(computedHmacHash, givenHmacHash)) {
            throw new RuntimeException("HMAC hashes do not match, make sure you used a matching .bckey file and password");
        }

        byte[] result = AESHelper.decryptData(privateKeyBytes, cryptoKey, IVec, false);

        System.out.println("AES decryption finished");
        return result;
    }

    public static byte[] decryptFile(
            final String encryptedFilePath,
            final byte[] fileCryptoKey,
            final String baseIVec,
            final int blockSize,
            final int offset,
            final int padding) throws UnsupportedEncodingException {
        System.out.println("AES decryption of file '" + encryptedFilePath + "' started");

        if (fileCryptoKey.length <= 0 || blockSize <= 0) {
            throw new RuntimeException("Crypto key for file can't be empty and block size must be bigger than zero");
        }

        // read the encrypted file
        byte[] fileBytes;
        try {
            fileBytes = Files.readAllBytes(Paths.get(encryptedFilePath));
        } catch (IOException e) {
            throw new RuntimeException("Could not read file", e);
        }

        // IVec in file header is base 64 encoded
        byte[] decodedFileIV = Base64Helper.decode(baseIVec);

        // report initial status
        int fileSize = fileBytes.length;
        long fileSizeFivePer = Double.valueOf(Math.floor(fileSize * 0.05)).longValue();
        String byteProgress = " (0 / " + String.valueOf(fileSize) + " bytes)";
        StringBuilder routeString = new StringBuilder();
        StringBuilder spaceString = new StringBuilder();
        IntStream.range(0, 20).forEach(i -> spaceString.append(" "));
        System.out.printf("Progress: [%s]%s%n", spaceString, byteProgress);

        // decrypt each block seperately with its own initialization vector
        int blockNo = 0;
        byte[] result = new byte[fileSize - offset - padding];
        for (int byteNo = offset, nextStatusThreshold = offset, currentStep = 0; byteNo < fileSize; byteNo += blockSize, ++blockNo) {
            byte[] blockIVec = AESHelper.computeBlockIVec(decodedFileIV, blockNo, fileCryptoKey);

            // get the input data for the current block (the last block may be shorter than [blockSize] bytes)
            int end = (byteNo + blockSize >= fileSize) ? fileSize : byteNo + blockSize;
            byte[] blockInput = Arrays.copyOfRange(fileBytes, byteNo, end);

            // PKCS7 padding for the last block if a cipher padding size greater than 0 was specified in file header
            // Note: the only differnce between PKCS5 and 7 is the block size (8 and 0-255 bytes respectively),
            // Java only offers the 'PKCS5PADDING' identifier (legacy from the time only 8 byte block ciphers were available)
            String currentPadding = (end == fileSize && padding > 0) ? "PKCS5PADDING" : "NOPADDING";

            // get the decrypted data for this block ...
            byte[] decryptedBlock = AESHelper.decryptData(blockInput, fileCryptoKey, blockIVec, true, currentPadding);

            // ... and append it to the previous data
            System.arraycopy(decryptedBlock, 0, result, byteNo - offset, decryptedBlock.length);

            // report intermediate status every 5%
            if (byteNo > nextStatusThreshold) {
                int steps = byteNo / nextStatusThreshold;
                nextStatusThreshold += fileSizeFivePer * steps;

                currentStep += steps;
                byteProgress = " (" + String.valueOf(byteNo) + " / " + String.valueOf(fileSize) + " bytes)";
                routeString.setLength(0);
                IntStream.range(0, currentStep).forEach(i -> routeString.append("#"));
                spaceString.setLength(0);
                IntStream.range(0, 20 - currentStep).forEach(i -> spaceString.append(" "));
                System.out.printf("Progress: [%s%s]%s%n", routeString, spaceString, byteProgress);
            }
        }

        // newline after status report
        byteProgress = " (" + String.valueOf(fileSize) + " / " + String.valueOf(fileSize) + " bytes)";
        routeString.setLength(0);
        IntStream.range(0, 20).forEach(i -> routeString.append("#"));
        System.out.printf("Progress: [%s]%s%n", routeString, byteProgress);

        System.out.println("AES decryption of file finished");
        return result;
    }
}
